### ISSUE TITLE
big-iq license manager templates fail onboarding

### ISSUE THREAD

**Issue by JeffGiroux** [F5Networks/f5-azure-arm-templates#161]
## Do you already have an issue opened with F5 support?
no

## Description
Deployed the standalone big-iq 2nic template. It fails at licensing. When I check logs, onboard.log shows the following main error "connect ECONNREFUSED 127.0.0.1:443 tries left: 90".

--snippet of logs
[azureuser@bigiq1:NO LICENSE:Standalone] azure # cat onboard.log 
2020-03-12T00:43:57.376Z info: [pid: 11234] [scripts/onboard.js] /config/cloud/azure/node_modules/@f5devcentral/f5-cloud-libs/scripts/onboard.js called with /usr/bin/node /config/cloud/azure/node_modules/@f5devcentral/f5-cloud-libs/scripts/onboard.js --host localhost --log-level debug --output /var/log/cloud/azure/onboard.log --signal ONBOARD_DONE --hostname f5vm01.westus.cloudapp.azure.com --cloud azure --license xxxxxx-xxxxxx-xxx-xxxxxx-xxxxxx --ntp 0.pool.ntp.org --tz UTC --create-reg-key-pool vlab-pool:xxxxxxxxxxx --dns 168.63.129.16 --big-iq-password-data-uri file:///mnt/cloudTmp/.bigiq_pass --metrics cloudName:azure,region:westus,customerId:26bcb58b0e5dbdcbae6683952de3bbc4efcd5315578f577ebff5b1a3604144a11f53433d37fcb441d2efb243871e647e0c1906829c220751fc9da2d4d60e0716,deploymentId:2678959d84a2ac812af2657842c502376ff2612d0cddd63965ca7f49a1942ce5f80505d009c53436fe82e318cc9006641f7d0b433b86c692e17c7e6c7892a265,licenseType:byol,bigIpVersion:6.0.0,templateName:f5-existing-stack-byol-2nic-bigiq,templateVersion:4.3.0 
2020-03-12T00:43:57.386Z info: [pid: 11234] [scripts/onboard.js] Onboard starting. 
2020-03-12T00:43:57.386Z debug: [pid: 11234] [lib/ipc.js] Sending signal ONBOARD_RUNNING 
2020-03-12T00:43:57.390Z info: [pid: 11234] [scripts/onboard.js] Generating temporary user. 
2020-03-12T00:44:00.962Z info: [pid: 11234] [scripts/onboard.js] Initializing device. 
2020-03-12T00:44:00.998Z info: [pid: 11234] [lib/bigIp.js] This is a BIG-IQ 
2020-03-12T00:44:00.999Z debug: [pid: 11234] [lib/authn.js] Getting auth token. 
2020-03-12T00:44:01.034Z verbose: [pid: 11234] [lib/util.js] tryUntil error: connect ECONNREFUSED 127.0.0.1:443 tries left: 90 
2020-03-12T00:44:11.059Z verbose: [pid: 11234] [lib/util.js] tryUntil error:
--some "502 Bad Gateway" html output--

Then...it runs out of tries and eventually fails...

2020-03-12T00:59:51.309Z verbose: [pid: 11234] [lib/util.js] tryUntil error: tryUntil: max tries reached: Public URI path not registered tries left: 1 
2020-03-12T01:00:01.312Z debug: [pid: 11234] [lib/bigIp.js] list localhost /shared/echo/available 
2020-03-12T01:00:01.317Z debug: [pid: 11234] [lib/bigIp.js]  
2020-03-12T01:00:01.317Z debug: [pid: 11234] [lib/bigIp.js] list localhost /shared/identified-devices/config/device-info/available 
2020-03-12T01:00:01.321Z debug: [pid: 11234] [lib/bigIp.js]  
2020-03-12T01:00:01.324Z debug: [pid: 11234] [lib/bigIp.js] list localhost /shared/iapp/package-management-tasks/available 
2020-03-12T01:00:01.327Z verbose: [pid: 11234] [lib/util.js] tryUntil error: Public URI path not registered tries left: 0 
2020-03-12T01:00:01.328Z info: [pid: 11234] [lib/bigIp.js] Device initialization failed tryUntil: max tries reached: tryUntil: max tries reached: Public URI path not registered 
2020-03-12T01:00:01.329Z debug: [pid: 11234] [lib/ipc.js] Sending signal CLOUD_LIBS_ERROR 
2020-03-12T01:00:01.330Z debug: [pid: 11234] [scripts/onboard.js]  
2020-03-12T01:00:01.327Z verbose: [pid: 11234] [lib/util.js] Max tries reached. 
2020-03-12T01:00:01.328Z verbose: [pid: 11234] [lib/util.js] tryUntil error: tryUntil: max tries reached: Public URI path not registered tries left: 0 
2020-03-12T01:00:01.328Z verbose: [pid: 11234] [lib/util.js] Max tries reached. 
2020-03-12T01:00:01.330Z error: [pid: 11234] [lib/util.js] Onboard failed: tryUntil: max tries reached: tryUntil: max tries reached: Public URI path not registered 
2020-03-12T01:00:01.330Z info: [pid: 11234] [scripts/onboard.js] Deleting temporary user. 
2020-03-12T01:00:02.064Z debug: [pid: 11234] [lib/ipc.js] Sending signal ONBOARD_DONE 
2020-03-12T01:00:02.064Z info: [pid: 11234] [lib/util.js] ERROR signaled from other script. Exiting 

cat cloudLibsError.log 
{"level":"error","message":"Onboard failed: tryUntil: max tries reached: tryUntil: max tries reached: Public URI path not registered","label":"lib/bigIpOnboard.js","timestamp":"2020-03-12T01:00:01.329Z"}

## Template
https://github.com/F5Networks/f5-azure-arm-templates/tree/master/experimental/bigiq/licenseManagement/standalone/2nic/existing-stack/byol

## Severity Level
3

**Comment by JeffGiroux** [comment]
Tested a new deployment. Still fails licensing. When I tested this time, I did not see the connect ECONNREFUSED 127.0.0.1:443 tries left but all remaining errors look identical to my first run.

azure # more onboard.log 
2020-03-12T01:27:36.213Z info: [pid: 14339] [scripts/onboard.js] /config/cloud/azure/node_modules/@f5devcentral/f5-cloud-libs/scripts/onboar
d.js called with /usr/bin/node /config/cloud/azure/node_modules/@f5devcentral/f5-cloud-libs/scripts/onboard.js --host localhost --log-level 
debug --output /var/log/cloud/azure/onboard.log --signal ONBOARD_DONE --hostname bigvm4.westus.cloudapp.azure.com --cloud azure --license BS
AVH-HTKNQT-BLX-VPPAMGD-DOEWAEF --ntp 0.pool.ntp.org --tz UTC --create-reg-key-pool vlab-pool:PAEQG-VVNFQ-CVDQX-MGVMY-NZZEFTF --dns 168.63.12
9.16 --big-iq-password-data-uri file:///mnt/cloudTmp/.bigiq_pass --metrics cloudName:azure,region:westus,customerId:26bcb58b0e5dbdcbae668395
2de3bbc4efcd5315578f577ebff5b1a3604144a11f53433d37fcb441d2efb243871e647e0c1906829c220751fc9da2d4d60e0716,deploymentId:98db53b3dc2e8da348756e
81dcad1723604967f57098c46861d6ea4e104a7d736f61efbd0acdc7b811c1fa213ca3a1f3629eb547502bce9f28449afd08936365,licenseType:byol,bigIpVersion:6.0
.0,templateName:f5-existing-stack-byol-2nic-bigiq,templateVersion:4.3.0 
2020-03-12T01:27:36.222Z info: [pid: 14339] [scripts/onboard.js] Onboard starting. 
2020-03-12T01:27:36.222Z debug: [pid: 14339] [lib/ipc.js] Sending signal ONBOARD_RUNNING 
2020-03-12T01:27:36.250Z info: [pid: 14339] [scripts/onboard.js] Generating temporary user. 
2020-03-12T01:27:37.525Z info: [pid: 14339] [scripts/onboard.js] Initializing device. 
2020-03-12T01:27:37.531Z info: [pid: 14339] [lib/bigIp.js] This is a BIG-IQ 
2020-03-12T01:27:37.533Z debug: [pid: 14339] [lib/authn.js] Getting auth token. 
2020-03-12T01:27:38.033Z verbose: [pid: 14339] [lib/util.js] tryUntil error:
`<html>
<head><title>502 Bad Gateway</title></head>
<body bgcolor="white">
<center><h1>502 Bad Gateway</h1></center>
<hr><center>webd</center>
</body>
</html>`
 tries left: 90

**Comment by JeffGiroux** [comment]
I'm deploying a BIG-IQ cluster template too. I'm still having issues getting both instances to launch, but I have at least one of the clustered instances of BIG-IQ running. It too fails licensing and/or onboarding. Same section in logs for lib/util.js and then the 502 bad gateway. I have a qkview of both the standalone device and the one success clustered device if needed.

**Comment by JeffGiroux** [comment]
Of note, I have tried BIG-IQ version 7.0 and it too fails. I have realized that the 502 bad gateway is a slight hiccup and the weird ERR connection refused 403 to 127.0.0.1 is also a random hiccup...unrelated to failure (I think). 

So far...
1. I have tested the AWS single BIG-IQ LM template. It works
2. I have tested the AWS cluster BIG-IQ LM template. It works.
3. I have tested both Azure BIG-IQ single and cluster. They launch the VMs, but onboarding fails.

Based on log differences in AWS and Azure, here's the onboarding log where things look different and start to fail for Azure.

AWS (good)
2020-03-13T18:27:40.043Z verbose: [pid: 11424] [lib/util.js] tryUntil error: tryUntil: max tries reached: not authorized tries left: 86 
2020-03-13T18:27:50.044Z debug: [pid: 11424] [lib/bigIp.js] list localhost /shared/echo/available 
2020-03-13T18:27:50.050Z verbose: [pid: 11424] [lib/util.js] tryUntil error: not authorized tries left: 0 
2020-03-13T18:27:50.050Z verbose: [pid: 11424] [lib/util.js] Max tries reached. 
2020-03-13T18:27:50.050Z verbose: [pid: 11424] [lib/util.js] tryUntil error: tryUntil: max tries reached: not authorized tries left: 85 
2020-03-13T18:28:00.051Z debug: [pid: 11424] [lib/bigIp.js] list localhost /shared/echo/available 
2020-03-13T18:28:00.059Z debug: [pid: 11424] [lib/bigIp.js]  
2020-03-13T18:28:00.061Z debug: [pid: 11424] [lib/bigIp.js] list localhost /shared/identified-devices/config/device-info/available 
2020-03-13T18:28:00.068Z debug: [pid: 11424] [lib/bigIp.js]  
2020-03-13T18:28:00.069Z debug: [pid: 11424] [lib/bigIp.js] list localhost /tm/sys/available 
2020-03-13T18:28:00.075Z debug: [pid: 11424] [lib/bigIp.js]  
2020-03-13T18:28:00.076Z debug: [pid: 11424] [lib/bigIp.js] list localhost /tm/cm/available 
2020-03-13T18:28:00.081Z debug: [pid: 11424] [lib/bigIp.js]  
2020-03-13T18:28:00.083Z debug: [pid: 11424] [lib/bigIp.js] list localhost /tm/sys/mcp-state/ 
2020-03-13T18:28:00.103Z debug: [pid: 11424] [lib/bigIp.js]  {"kind":"tm:sys:mcp-state:mcp-statestats","selfLink":"https://localhost/mgmt/tm/sys/mcp-state?ver=12.1.1","entries":{"https://localhost/mgmt/tm/sys/mcp-state/0":{"nestedStats":{"entries":{"lastLoad":{"description":"full-config-load-succeed"},"phase":{"description":"running"}}}}}}
2020-03-13T18:28:00.103Z info: [pid: 11424] [scripts/onboard.js] Waiting for device to be ready. 
2020-03-13T18:28:00.104Z debug: [pid: 11424] [lib/bigIp.js] list localhost /shared/echo/available 
2020-03-13T18:28:00.110Z debug: [pid: 11424] [lib/bigIp.js]  
2020-03-13T18:28:00.111Z debug: [pid: 11424] [lib/bigIp.js] list localhost /shared/identified-devices/config/device-info/available 

Azure (bad)
2020-03-12T00:45:39.209Z verbose: [pid: 11234] [lib/util.js] tryUntil error: tryUntil: max tries reached: not authorized tries left: 86 
2020-03-12T00:45:49.212Z debug: [pid: 11234] [lib/bigIp.js] list localhost /shared/echo/available 
2020-03-12T00:45:49.218Z verbose: [pid: 11234] [lib/util.js] tryUntil error: not authorized tries left: 0 
2020-03-12T00:45:49.218Z verbose: [pid: 11234] [lib/util.js] Max tries reached. 
2020-03-12T00:45:49.220Z verbose: [pid: 11234] [lib/util.js] tryUntil error: tryUntil: max tries reached: not authorized tries left: 85 
2020-03-12T00:45:59.223Z debug: [pid: 11234] [lib/bigIp.js] list localhost /shared/echo/available 
2020-03-12T00:45:59.230Z debug: [pid: 11234] [lib/bigIp.js]  
2020-03-12T00:45:59.230Z debug: [pid: 11234] [lib/bigIp.js] list localhost /shared/identified-devices/config/device-info/available 
2020-03-12T00:45:59.235Z debug: [pid: 11234] [lib/bigIp.js]  
2020-03-12T00:45:59.239Z debug: [pid: 11234] [lib/bigIp.js] list localhost /shared/iapp/package-management-tasks/available 
2020-03-12T00:45:59.245Z verbose: [pid: 11234] [lib/util.js] tryUntil error: Public URI path not registered tries left: 0 
2020-03-12T00:45:59.245Z verbose: [pid: 11234] [lib/util.js] Max tries reached. 
2020-03-12T00:45:59.245Z verbose: [pid: 11234] [lib/util.js] tryUntil error: tryUntil: max tries reached: Public URI path not registered tries left: 84 
2020-03-12T00:46:09.249Z debug: [pid: 11234] [lib/bigIp.js] list localhost /shared/echo/available 
2020-03-12T00:46:09.255Z debug: [pid: 11234] [lib/bigIp.js]  
2020-03-12T00:46:09.259Z debug: [pid: 11234] [lib/bigIp.js] list localhost /shared/identified-devices/config/device-info/available 
2020-03-12T00:46:09.265Z debug: [pid: 11234] [lib/bigIp.js]  
2020-03-12T00:46:09.268Z debug: [pid: 11234] [lib/bigIp.js] list localhost /shared/iapp/package-management-tasks/available 
2020-03-12T00:46:09.273Z verbose: [pid: 11234] [lib/util.js] tryUntil error: Public URI path not registered tries left: 0 
2020-03-12T00:46:09.274Z verbose: [pid: 11234] [lib/util.js] Max tries reached. 
2020-03-12T00:46:09.274Z verbose: [pid: 11234] [lib/util.js] tryUntil error: tryUntil: max tries reached: Public URI path not registered tries left: 83 
2020-03-12T00:46:19.278Z debug: [pid: 11234] [lib/bigIp.js] list localhost /shared/echo/available 
2020-03-12T00:46:19.284Z debug: [pid: 11234] [lib/bigIp.js]  
2020-03-12T00:46:19.289Z debug: [pid: 11234] [lib/bigIp.js] list localhost /shared/identified-devices/config/device-info/available 
2020-03-12T00:46:19.294Z debug: [pid: 11234] [lib/bigIp.js]  
2020-03-12T00:46:19.298Z debug: [pid: 11234] [lib/bigIp.js] list localhost /shared/iapp/package-management-tasks/available 
2020-03-12T00:46:19.304Z verbose: [pid: 11234] [lib/util.js] tryUntil error: Public URI path not registered tries left: 0

**Comment by JeffGiroux** [comment]
Here is the lib/bigIp.js on the AWS (good) deploy.
/config/cloud/aws/node_modules/@f5devcentral/f5-cloud-libs/lib/bigIp.js

        const availabilityChecks = [
            '/shared/echo/available',
            '/shared/identified-devices/config/device-info/available',
            '/tm/sys/available',
            '/tm/cm/available'
        ];


Here is the Azure (bad)...notice it tries to call for iapps which fails
/config/cloud/azure/node_modules/@f5devcentral/f5-cloud-libs/lib/bigIp.js

--snippet--
        const availabilityChecks = [
            '/shared/echo/available',
            '/shared/identified-devices/config/device-info/available',
            '/shared/iapp/package-management-tasks/available',
            '/tm/sys/available',
            '/tm/cm/available'
        ];

**Comment by JeffGiroux** [comment]
I think this is due to the cloud-libs version. Check this...

Azure
"f5CloudLibsTag": "v4.16.0",
"f5CloudLibsAzureTag": "v2.13.0"

AWS
--fcl-tag v4.8.2 --fcl-cloud-tag v2.4.0 

Notice the old versions for AWS. I checked f5-cloud-libs version v4.8.2 and it does NOT include a check for iapps. This would be why AWS succeeds and Azure fails as AWS does NOT check for iapp and hence does not fail on the broken command. The version 4.16.0 f5-cloud-libs that Azure pulls down does have a check for iapps. I scrubbed versions, and it appears the iapp check was added in tag 4.10.3 of f5-cloud-libs. Link below.

https://github.com/F5Networks/f5-cloud-libs/blob/v4.10.3/lib/bigIp.js#L989
